cmasher/utils.py

```diff
@@ -233,6 +239,100 @@ def _get_cmap_perceptual_rank(
+
+    Parameters:
+        cmaps (list[Union[Colormap, str]]): list of colormaps to combine.
+        nodes (List[float], optional): The blending points between
+            colormaps, in the range [0, 1]. Defaults to equal divisions.
+        n_rgb_levels (int, optional): Number of RGB levels for each colormap
+            segment. Defaults to 256.
+
+    Returns:
+        Colormap: The composite colormap.
+
+    Raises:
+        TypeError: If the list contains mixed datatypes or invalid
+            colormap names.
+        ValueError: If the cmaps contain only one single colormap,
+            or if the number of nodes is not one less than the number
+            of colormaps, or if the nodes do not contain incrementing values
+            between 0.0 and 1.0.
+
+    Note:
+        The colormaps are combined from bottom to top.
+
+    References:
+        - https://stackoverflow.com/questions/31051488/combining-two-matplotlib-colormaps/31052741#31052741
+
+    Examples:
+        # Example 1: Using colormap names
+        >>> cmaps = ["ocean", "prism", "coolwarm"]
+        >>> nodes = [0.2, 0.75]
+        >>> custom_cmap_1 = combine_cmaps(cmaps, nodes)
+
+        # Example 2: Using predefined colormaps
+        >>> cmap_0 = plt.get_cmap("Blues")
+        >>> cmap_1 = plt.get_cmap("Oranges")
+        >>> cmap_2 = plt.get_cmap("Greens")
+        >>> custom_cmap_2 = combine_cmaps([cmap_0, cmap_1, cmap_2])
+
+    """
+    # Check colormap datatype and convert to list[Colormap]
+    if len(cmaps) == 1:
+        raise ValueError("Cannot combine a single colormap.")
+    for cm in cmaps:
+        if not isinstance(cm, (Colormap, str)):
+            raise TypeError(f"Unsupported colormap type: {type(cm)}.")
+    cmaps = [cm if isinstance(cm, Colormap) else mpl.colormaps[cm] for cm in cmaps]
+
+    # Generate default nodes for equal separation
+    if nodes is None:
+        nodes = np.linspace(0, 1, len(cmaps) + 1)[1:-1].tolist()
+
+    # Check nodes length
+    if len(nodes) != len(cmaps) - 1:
+        raise ValueError(
+            "Number of nodes should be one less than the number of colormaps."
+        )
+
+    # Preprocess nodes list for easier segmentation
+    nodes.insert(0, 0.0)
+    nodes.append(1.0)
+
+    # Check node values
+    if not all(0.0 <= node <= 1.0 for node in nodes) or any(
+        nodes[i] >= nodes[i + 1] for i in range(len(nodes) - 1)
+    ):
+        raise ValueError(
+            "Nodes should contain incrementing values between 0.0 and 1.0."
+        )
+
+    # Generate composite colormap
+    combined_cmap_segments = []
+
+    for i, cmap in enumerate(cmaps):
+        start_position = nodes[i]
+        end_position = nodes[i + 1]
+
+        # Calculate the length of the segment
+        segment_length = int(n_rgb_levels * (end_position - start_position))
+
+        # Append the segment to the combined colormap segments
+        combined_cmap_segments.append(cmap(np.linspace(0, 1, segment_length)))
```

Comment:
It appears `mypy` is expecting `cmap` to be `str`, and therefore:
```
> Run mypy cmasher
cmasher/utils.py: note: In function "combine_cmaps":
cmasher/utils.py:328: error: "str" not callable  [operator]
Found 1 error in 1 file (checked 63 source files)
Error: Process completed with exit code 1.
```

As I wish to maintain compatibility with both `str` and `Colormap` by using `list[Union[Colormap, str]]`, instead of forcing `cmap` to be `Colormap` altogether. Can you please give me some advice on this type annotation here? @neutrinoceros Thanks!

Reply:
What's happening is that you implemented logic that narrows the type of `cmaps` from `list[Colormap | str]` to `list[Colomap]` (which is great), but mypy isn't able to understand that logic, so it still assumes `list[Colormap | str]` at this point, even though we, as humans, know that the type is really `list[Colomap]` here.

A solution is to use a second variable (with its own type annotation) instead of rebinding the input parameter. For instance, you'd do
```python
cmaps_: list[Colormap] = [cm if isinstance(cm, Colormap) else mpl.colormaps[cm] for cm in cmaps]
```
and to use `cmaps_` internally instead of `cmaps`.

Reply:
Hi @neutrinoceros thanks a lot for the quick response and explaination. Yes actually the type narrows from `list[Colormap | str]` to `list[Colormap]` during the run time thus the error. 

And I also tried your suggestion to initiate a private variable and static typed `_cmap` which seems to resolve the `mypy` error. Meanwhile I'm a little bit uncertain about whether we should strictly follow suggestions from `mypy`, as it seems to be making our code statically typed (though Python should be dynamically typed)? Though I have to admit using static typing does in many cases help avoid potential typing issues. Should I use `# type: ignore` to suppress the error?

Reply:
Please refrain from ignoring type checking errors when there's a clear solution. Statically typed code is much easier to reason about and to maintain; using strict types for internal variables doesn't force us to renounce flexibility in the user interface, which is where Python's dynamic behavior is really useful.